src/caches/beliefpropagationcache.jl

```diff
@@ -106,7 +106,8 @@ end
 
 function message(bp_cache::BeliefPropagationCache, edge::PartitionEdge)
   mts = messages(bp_cache)
-  return get(mts, edge, default_message(bp_cache, edge))
+  haskey(mts, edge) && return mts[edge]
+  return default_message(bp_cache, edge)
```

Comment:
Is the issue basically that `default_message(bp_cache, edge)` always gets evaluated, even if the message already exists? There is a version of `get` that accepts a function as the first argument, which makes it so that the default is only evaluated if needed:
```julia
julia> get(Dict("x" => 2), "x", println("test"))
test
2

julia> get(Dict("x" => 2), "x") do
         println("test")
       end
2
```

Reply:
Yeah it seems to be evaluating `default` always which is a lot of unnecessary allocation and causes a factor 2/3 slowdown in some of my tests. Didn't realize that if you made the default the first argument and a function it would fix it. I will do that.